File: Task/Dutch-national-flag-problem/Ruby/dutch-national-flag-problem.rb
class Ball
  FLAG = {red: 1, white: 2, blue: 3}

  def initialize
    @color = FLAG.keys.sample
  end

  def color
    @color
  end

  def <=>(other)  # needed for sort, results in -1 for <, 0 for == and 1 for >.
    FLAG[self.color] <=> FLAG[other.color]
  end

  def inspect
    @color
  end
end

balls = []
balls = Array.new(8){Ball.new} while balls == balls.sort

puts "Random: #{balls}"
puts "Sorted: #{balls.sort}"

File: Task/Dutch-national-flag-problem/Clojure/dutch-national-flag-problem.clj
(defn dutch-flag-order [color]
  (get {:red 1 :white 2 :blue 3} color))

(defn sort-in-dutch-flag-order [balls]
  (sort-by dutch-flag-order balls))

;; Get a collection of 'n' balls of Dutch-flag colors
(defn random-balls [num-balls]
  (repeatedly num-balls
              #(rand-nth [:red :white :blue])))

;; Get random set of balls and insure they're not accidentally sorted
(defn starting-balls [num-balls]
  (let [balls (random-balls num-balls)
        in-dutch-flag-order? (= balls
                                (sort-in-dutch-flag-order balls))]
    (if in-dutch-flag-order?
      (recur num-balls)
      balls)))
